storage: introduce indirection between lock table keys and lock strength

arulajmani opened the pull request:
Lock table keys include a single byte that denotes the lock's strength. Prior to this patch, that byte was directly derived from the lock strength. Versions <= 23.1 considered intents to have Exclusive lock strength; versions > 23.1 treat intents as having a new (stronger) lock strength -- lock.Intent.

This requires us to change how lock table keys are serialized and deserialized -- otherwise, we'd have a incompatibility issue between intents written across these versions. To that end, this patch introduces a layer of indirection between a lock's strength and the strength byte that's persisted in lock table keys. We then ensure the byte itself doesn't change between versions -- i.e, all versions use 3 (the enum value of `lock.Exclusive`) as the byte written by an intent.

Closes #107947

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/108136)
<!-- Reviewable:end -->

nvb reviewed (commented):
> Reviewed 3 of 10 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108136)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @RaduBerinde)

___
*[`pkg/storage/engine_key.go` line 273 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108136#-Nax2tCa91mw-o2tqe4Z:-Nax2tCa91mw-o2tqe4_:bc83va1) ([raw file](https://github.com/cockroachdb/cockroach/blob/83818785acf9094d81679526942e6d9b84b040ca/pkg/storage/engine_key.go#L273)):*
> ```Go
> // (lock.Strength, StrengthByte), where the StrengthByte corresponds to the byte
> // persisted in LockTableKeys.
> var replicatedLockStrengthToByteMap = func() [lock.NumLockStrength]byte {
> ```

This is close, but the way this usually looks in Go is a little bit more like:
```go
var strToByte = [...]byte{
	lock.Shared:    1,
	lock.Exclusive: 2,
	lock.Intent:    3,
}

var byteToStr = func() (arr [len(strToByte) + 1]lock.Strength) {
	for str, b := range strToByte {
		arr[b] = str
	}
	return arr
}()

func getByteForStr(str int) (byte, error) {
	if str < 0 || str >= len(strToByte) {
		return 0, errors.Error("...")
	}
	b := strToByte[str]
	if b == 0 {
		return 0, errors.Error("...")
	}
	return b, nil
}

func getStrForByte(b byte) (lock.Strength, error) {
	// ... same thing. Maybe try generics.
}
```

You could even strongly type `byte` for more type safety. And add a `mustGetByteForStr` function for testing.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108136)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @RaduBerinde)

___
*[`pkg/storage/engine_key.go` line 266 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108136#-NbGHVs_3UD8_dSp18CC:-NbGHVs_3UD8_dSp18CD:b-jd4514) ([raw file](https://github.com/cockroachdb/cockroach/blob/83818785acf9094d81679526942e6d9b84b040ca/pkg/storage/engine_key.go#L266)):*
> ```Go
> var byteToReplicatedLockStrengthMap = func() [lock.NumLockStrength]lock.Strength {
> 	var m [lock.NumLockStrength]lock.Strength
> 	m[3] = lock.Intent
> ```

If both `lock.Exclusive` and `lock.Intent` are using the same byte 3, how can one map from byte to `lock.Strength`?

Are we never persisting `lock.Exclusive` locks?


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108136)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @RaduBerinde, and @sumeerbhola)

___
*[`pkg/storage/engine_key.go` line 266 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108136#-NbGHVs_3UD8_dSp18CC:-Nbzjobg2-MDEhaMHbq5:b9pz356) ([raw file](https://github.com/cockroachdb/cockroach/blob/83818785acf9094d81679526942e6d9b84b040ca/pkg/storage/engine_key.go#L266)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

If both `lock.Exclusive` and `lock.Intent` are using the same byte 3, how can one map from byte to `lock.Strength`?

Are we never persisting `lock.Exclusive` locks?

</blockquote></details>

The intention is for `lock.Exclusive` to map to/from byte 2. This means that all prior encoded keys will now be decoded as `lock.Intent`. Thus far, we have never persisted `lock.Exclusive` locks.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
This PR is going to need to change a bit with https://github.com/cockroachdb/cockroach/pull/109082.

I also realized while typing that refactor out that we probably don't want to change the `LockTableKey.Strength` field itself. Instead, we can push the translation directly into `LockTableKey.ToEngineKey` and `EngineKey.ToLockTableKey`. I suspect that will make this change much smaller by hiding the translation behind the `LockTableKey` abstraction.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108136#-:-NcDoOl6CqxZ9apH4OFH:b-jqjikw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @RaduBerinde, and @sumeerbhola)



<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> Instead, we can push the translation directly into `LockTableKey.ToEngineKey` and `EngineKey.ToLockTableKey`.

Ack, I'll make the change. Haven't looked at the other patch yet, but do you want me to build this on top of yours? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108136#-:-NcDqpzBDPsXZ-2w_wAO:b-sn4t5v)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @sumeerbhola)



<!-- Sent from Reviewable.io -->

arulajmani committed:
storage: introduce indirection between lock table keys and lock strength

Lock table keys include a single byte that denotes the lock's strength.
Prior to this patch, that byte was directly derived from the lock
strength. Versions <= 23.1 considered intents to have Exclusive lock
strength; versions > 23.1 treat intents as having a new (stronger) lock
strength -- lock.Intent.

This requires us to change how lock table keys are serialized and
deserialized -- otherwise, we'd have a incompatibility issue between
intents written across these versions. To that end, this patch
introduces a layer of indirection between a lock's strength and the
strength byte that's persisted in lock table keys. We then ensure the
byte itself doesn't change between versions -- i.e, all versions use 3
(the enum value of lock.Exclusive) as the byte written by an intent.

Note that until 23.1, the only replicated locks we supported were
intents, so this performing this remapping is copacetic. In the future
(23.2) when we support replicated locks of strength `Exclusive`, we'll
use this layer of indirection to assign them a different strength byte
(say 2).

Closes #107947
Release note: None

arulajmani reviewed (commented):
Rebased this + took your suggestion to push the translation into `LockTableKey.ToEngineKey` and `EngineKey.ToLockTableKey`. RFAL.

After coming back to this and implementing your suggestion, I'm wondering how you feel about renaming `LockTableKey` to `ReplicatedLock`? It helps make it clear that the `Strength` field has nothing to do with the key encoding. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108136#-:-NcTM1OR4Cg4Qp_PvHTG:baz1pkk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @RaduBerinde, and @sumeerbhola)

___
*[`pkg/storage/engine_key.go` line 266 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108136#-NbGHVs_3UD8_dSp18CC:-NcTNvpC2x0s3TeTdzU9:baf8829) ([raw file](https://github.com/cockroachdb/cockroach/blob/83818785acf9094d81679526942e6d9b84b040ca/pkg/storage/engine_key.go#L266)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

The intention is for `lock.Exclusive` to map to/from byte 2. This means that all prior encoded keys will now be decoded as `lock.Intent`. Thus far, we have never persisted `lock.Exclusive` locks.
</blockquote></details>

I added some words about this in the commit message to clarify.
___
*[`pkg/storage/engine_key.go` line 273 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108136#-Nax2tCa91mw-o2tqe4Z:-NcTKjoMCnItsXzOiX40:bszo8gw) ([raw file](https://github.com/cockroachdb/cockroach/blob/83818785acf9094d81679526942e6d9b84b040ca/pkg/storage/engine_key.go#L273)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This is close, but the way this usually looks in Go is a little bit more like:
```go
var strToByte = [...]byte{
	lock.Shared:    1,
	lock.Exclusive: 2,
	lock.Intent:    3,
}

var byteToStr = func() (arr [len(strToByte) + 1]lock.Strength) {
	for str, b := range strToByte {
		arr[b] = str
	}
	return arr
}()

func getByteForStr(str int) (byte, error) {
	if str < 0 || str >= len(strToByte) {
		return 0, errors.Error("...")
	}
	b := strToByte[str]
	if b == 0 {
		return 0, errors.Error("...")
	}
	return b, nil
}

func getStrForByte(b byte) (lock.Strength, error) {
	// ... same thing. Maybe try generics.
}
```

You could even strongly type `byte` for more type safety. And add a `mustGetByteForStr` function for testing.
</blockquote></details>

Thanks for sketching this out! I took ~most of your suggestions. I didn't add `lock.Shared` or `lock.Exclusive` to the map just yet.

Now that we're no longer storing `StrengthByte` in the `LockTableKey` struct, I didn't feel strongly typing `byte` would get us much. Happy to iterate over this.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> After coming back to this and implementing your suggestion, I'm wondering how you feel about renaming `LockTableKey` to `ReplicatedLock`? It helps make it clear that the `Strength` field has nothing to do with the key encoding.

Discussed on slack and decided to keep this as `LockTableKey`.

> Reviewed 12 of 12 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108136#-:-NcTVm4b6UqX5BWJZMmL:bog2h9z)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @RaduBerinde, and @sumeerbhola)

___
*[`pkg/storage/engine_key.go` line 273 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108136#-Nax2tCa91mw-o2tqe4Z:-NcTwUqh0h_qamEuJRs3:b9jrqxv) ([raw file](https://github.com/cockroachdb/cockroach/blob/83818785acf9094d81679526942e6d9b84b040ca/pkg/storage/engine_key.go#L273)):*
> I didn't feel strongly typing `byte` would get us much. Happy to iterate over this.

Agreed. I don't think it's needed now.
___
*[`pkg/storage/engine_key.go` line 258 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108136#-NcTwb8M8reff1iuAO53:-NcTwb8M8reff1iuAO54:b-wlfqbi) ([raw file](https://github.com/cockroachdb/cockroach/blob/da51833a932c2ab1453c28029aada86c7fdc58d8/pkg/storage/engine_key.go#L258)):*
> ```Go
> }
> 
> // byteToReplicatedLockStrength is a mapping between the strength byte persisted
> ```

small nit: move this below `replicatedLockStrengthToByte`, since this is derived from that array.
___
*[`pkg/storage/engine_key.go` line 261 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108136#-NcV1ANV3oePsWjIaEcU:-NcV1ANV3oePsWjIaEcV:b6xsz7n) ([raw file](https://github.com/cockroachdb/cockroach/blob/da51833a932c2ab1453c28029aada86c7fdc58d8/pkg/storage/engine_key.go#L261)):*
> [len(replicatedLockStrengthToByte)]

I thought this had to be `len(replicatedLockStrengthToByte)+1`, but what you have here also works. It's actually a bit subtle why though, having to do with `lock.Intent`'s value of 4. We should probably just make this a slice and be more explicit. Something like:
```go
var byteToReplicatedLockStrength = func() (arr []lock.Strength) {
    maxByte := 0
    for _, b := range replicatedLockStrengthToByte {
        if b > maxBytes {
            b = maxByte
        }
    }
    arr = make([]lock.Strength, maxByte+1)
    for str, b := range replicatedLockStrengthToByte {
        if b != 0 {
            arr[b] = lock.Strength(str)
        }
    }
    return arr
}()
```
___
*[`pkg/storage/engine_key.go` line 278 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108136#-NcTwkMOEBEHr0ngCATI:-NcTwkMOEBEHr0ngCATJ:b-xg56pd) ([raw file](https://github.com/cockroachdb/cockroach/blob/da51833a932c2ab1453c28029aada86c7fdc58d8/pkg/storage/engine_key.go#L278)):*
> ```Go
> // a lock's key encoding, given its lock strength.
> func getByteForReplicatedLockStrength(str lock.Strength) byte {
> 	if str <= lock.None || int(str) >= len(replicatedLockStrengthToByte) {
> ```

nit: `s/str <= lock.None/str < 0/`, since the intention is to use `str` as an index into the array.
___
*[`pkg/storage/engine_key.go` line 291 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108136#-NcV-SiTDOzohuaMtAIq:-NcV-SiTDOzohuaMtAIr:bd7r55i) ([raw file](https://github.com/cockroachdb/cockroach/blob/da51833a932c2ab1453c28029aada86c7fdc58d8/pkg/storage/engine_key.go#L291)):*
> ```Go
> // the strength byte from its key encoding.
> func getReplicatedLockStrengthForByte(b byte) (lock.Strength, error) {
> 	if b <= 0 || int(b) >= len(byteToReplicatedLockStrength) {
> ```

`s/<=/</`
___
*[`pkg/storage/engine_key.go` line 295 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108136#-NcV-km69MYe3xZOxjYQ:-NcV-km69MYe3xZOxjYR:b-w378ox) ([raw file](https://github.com/cockroachdb/cockroach/blob/da51833a932c2ab1453c28029aada86c7fdc58d8/pkg/storage/engine_key.go#L295)):*
> ```Go
> 	}
> 	str := byteToReplicatedLockStrength[b]
> 	if str <= lock.None || str > lock.Intent {
> ```

Should this just be `if str == 0 {`?


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108136#-:-NcZYiol4kNKJHzTeOTh:bb72pd0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @RaduBerinde, and @sumeerbhola)

___
*[`pkg/storage/engine_key.go` line 261 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108136#-NcV1ANV3oePsWjIaEcU:-NcZVPHfE97eUTrsdogh:b-woljjc) ([raw file](https://github.com/cockroachdb/cockroach/blob/da51833a932c2ab1453c28029aada86c7fdc58d8/pkg/storage/engine_key.go#L261)):*
> It's actually a bit subtle why though, having to do with `lock.Intent`'s value of 4.

I agree. I also like this bit: 

```
 if b != 0 { arr[b] = lock.Strength(str) {
```

it's much clearer than the old thing.
___
*[`pkg/storage/engine_key.go` line 291 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108136#-NcV-SiTDOzohuaMtAIq:-NcZY1GgCgRY_h8jOoQQ:bhz681) ([raw file](https://github.com/cockroachdb/cockroach/blob/da51833a932c2ab1453c28029aada86c7fdc58d8/pkg/storage/engine_key.go#L291)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`s/<=/</`
</blockquote></details>

We can get rid of this entirely, because a byte can never be < 0.
___
*[`pkg/storage/engine_key.go` line 295 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108136#-NcV-km69MYe3xZOxjYQ:-NcZWxpI7ehY-3Ij6PDj:bm0ujvn) ([raw file](https://github.com/cockroachdb/cockroach/blob/da51833a932c2ab1453c28029aada86c7fdc58d8/pkg/storage/engine_key.go#L295)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should this just be `if str == 0 {`?
</blockquote></details>

Yeah, `str == 0` makes more sense. 


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108136#-:-Nc_T3Wh7uNCnhZwTCYI:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @RaduBerinde and @sumeerbhola)



<!-- Sent from Reviewable.io -->

arulajmani commented:
TFTR! 

bors r=nvanbenschoten
